### PR TITLE
Adds test for caching result set with file engine

### PR DESCRIPTION
**Pull request by thinkingmedia**
Adds a test for verifying that ResultSet objects can be cached with the file engine, and deserialized again with the same data.

New unit test is failing for me on PHP 7.1.1

Relates to issue #10111

### CODE REVIEW

File: tests/TestCase/Cache/Engine/FileEngineTest.php [hunk 1]
```diff
@@ -193,6 +207,34 @@ public function testSerialize()
...
+
+        $this->_configCache(['serialize' => true]);
+        $write = Cache::write('articles', $query, 'file_test');
+        $this->assertTrue($write);
+
+        $read = Cache::read('articles', 'file_test');
+        $this->assertInstanceOf(ResultSet::class, $read);
+        $this->assertSame(3, count($read->toArray()));
+
+        TableRegistry::clear();
```
Comment (markstory) [inline tests/TestCase/Cache/Engine/FileEngineTest.php]: Couldn't this test be a test for `ResultSet` being serialized and unserialized as that is where the actual issue was.

File: tests/TestCase/ORM/EntityTest.php [hunk 2]
```diff
@@ -713,6 +713,20 @@ public function testJsonSerialize()
...
+     *
+     * @return void
+     */
+    public function testPhpSerialize()
+    {
+        $data = ['name' => 'James', 'age' => 20, 'phones' => ['123', '457']];
+        $entity = new Entity($data);
+        $copy = unserialize(serialize($entity));
+        $this->assertInstanceOf(Entity::class, $copy);
+        $this->assertEquals($data, $copy->toArray());
```
Comment (markstory) [inline tests/TestCase/ORM/EntityTest.php]: Is this adding coverage in general? It doesn't look related to the defect that is being fixed.

Comment (thinkingmedia) [inline tests/TestCase/ORM/EntityTest.php]: I wanted to know if the serializing issue was related to ResultSet or Entities. So this passed and I just kept it in.

File: tests/TestCase/Cache/Engine/FileEngineTest.php (hunk 1 above)
Comment (thinkingmedia) [inline tests/TestCase/Cache/Engine/FileEngineTest.php]: Yes, I can remove this test. It doesn't add anything.

File: tests/TestCase/ORM/EntityTest.php (hunk 2 above)
Comment (markstory) [inline tests/TestCase/ORM/EntityTest.php]: Ok.

**Comment by thinkingmedia** [comment]
Anyone think it should be this instead?

```
  $this->_results = (array)(unserialize($serialized) ?: []);
```

**Comment by markstory** [comment]
When would the unserialized value be falsey?

**Comment by thinkingmedia** [comment]
@markstory on failure the function returns false. If we don't use `?:` then the result would be `[false]`.

**Comment by thinkingmedia** [comment]
@markstory Don't merge. I'm able to reproduce the Travis errors locally. I'm investigating tonight or tomorrow.

File: src/ORM/ResultSet.php [hunk 3]
```diff
@@ -313,10 +313,19 @@ public function first()
...
+        if (!$this->_useBuffering) {
+            $msg = 'You cannot serialize an un-buffered ResultSet. Use Query::bufferResults() to get a buffered ResultSet.';
+            throw new Exception($msg);
+        }
+
         while ($this->valid()) {
             $this->next();
         }
 
+        if ($this->_results instanceof SplFixedArray) {
```
Comment (lorenzo) [inline src/ORM/ResultSet.php]: I'm considering just removing SplFixedArray... it did not perform as good as I wanted it to

**Comment by lorenzo** [comment]
Thanks! @thinkingmedia

File: src/ORM/ResultSet.php (hunk 3 above)
Comment (thinkingmedia) [inline src/ORM/ResultSet.php]: Why this broke now and not a year ago is very very strange. I don't think there was anything wrong with using SplFixedArray. There must have been some dependency change in PHP that was across 5.6 and up versions that triggered this. We should report this to the PHP team.
